### PR TITLE
[REVIEW] Fix hash map test failure

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Hash map test was failing due to missing `cudaStreamSynchronize(stream)`call, this was removed in the #3548 as per the discussion https://github.com/rapidsai/cudf/pull/3548#discussion_r355078040, so calling cuda sync in test.

### CODE REVIEW

**Comment by rgsl888prabhu** [comment]
> @rgsl888prabhu thanks for the fix. As part of this fix, would you be willing to add some documentation explaining a few things?
> 
> * In the class docs, add some explanation that the user is responsible for following stream semantics: either the same stream should be used to create the map as is used by the kernels that access it, or the stream should be synchronized before it is accessed from a different stream or on the host.
> * For `create()`, document that all allocations, kernels and copies in the constructor take place on `stream` but that it does not synchronize the stream.
> * Add docs for all the begin() and end() functions that specify that these can only be used on the host if the managed allocator is used, and that the caller should ensure the stream is synchronized before it is accessed from the host. If called from device code, it should either be running on the same stream as create(), or the accessing stream should be appropriately synchronized with the creating stream.

I will take care of it.

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 1]
```diff
@@ -86,6 +86,10 @@ union pair_packer<pair_type, std::enable_if_t<is_packable<pair_type>()>> {
 /**
  * Supports concurrent insert, but not concurrent insert and find.
  *
+ * @note User is responsible for following stream semantics:
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]:
```suggestion
 * @note The user is responsible for the following stream semantics:
```

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 2]
```diff
@@ -86,6 +86,10 @@ union pair_packer<pair_type, std::enable_if_t<is_packable<pair_type>()>> {
...
+ * - Either same stream should be used to create the map as is used by the kernels that access it
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]:
```suggestion
 * - Either the same stream should be used to create the map as is used by the kernels that access it, or
```

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 3]
```diff
@@ -86,6 +86,10 @@ union pair_packer<pair_type, std::enable_if_t<is_packable<pair_type>()>> {
...
+ * - Or stream should be synchronized before it is accessed from a different stream or on the host
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]:
```suggestion
 * - the stream used to create the map should be synchronized before it is accessed from a different stream or from host code.
```

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 4]
```diff
@@ -86,6 +86,10 @@ union pair_packer<pair_type, std::enable_if_t<is_packable<pair_type>()>> {
...
+ *
  * TODO:
  *  - add constructor that takes pointer to hash_table to avoid allocations
  *  - extend interface to accept streams
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Is this TODO about streams still needed?

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 5]
```diff
@@ -116,10 +120,14 @@ class concurrent_unordered_map {
...
    * @note The implementation of this unordered_map uses sentinel values to
    * indicate an entry in the hash table that is empty, i.e., if a hash bucket
-   *is empty, the pair residing there will be equal to (unused_key,
-   *unused_element). As a result, attempting to insert a key equal to
+   * is empty, the pair residing there will be equal to (unused_key,
+   * unused_element). As a result, attempting to insert a key equal to
    *`unused_key` results in undefined behavior.
    *
+   * @note All allocations, kernels and copies in the constructor take place
+   * on stream but that it does not synchronize the stream. It is user's
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]:
```suggestion
   * on stream but the constructor does not synchronize the stream. It is the user's
```

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 6]
```diff
@@ -116,10 +120,14 @@ class concurrent_unordered_map {
...
+   * responsibility to synchronize.
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]:
```suggestion
   * responsibility to synchronize or use the same stream to access the map.
```

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 7]
```diff
@@ -149,18 +157,53 @@ class concurrent_unordered_map {
         deleter};
   }
 
+  /**
+   * @note If called on host using managed allocator, then caller should ensure
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Add @brief. Also, this is a device-only function (unlike multi map), not sure why.

```suggestion
   * @brief Returns an iterator to the first element in the map
   *
```

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 8]
```diff
@@ -149,18 +157,53 @@ class concurrent_unordered_map {
...
+   * the stream is synchronized before it is accessed from the host.
+   *
+   * @note When called in a device code, user should make sure that it should
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]:
```suggestion
   * @note `__device__` code that calls this function should either run in the same stream as `create()`, or the accessing stream
```

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 9]
```diff
@@ -149,18 +157,53 @@ class concurrent_unordered_map {
...
+   * either be running on the same stream as create(), or the accessing stream
+   * should be appropriately synchronized with the creating stream.
+   **/
   __device__ iterator begin() {
     return iterator(m_hashtbl_values, m_hashtbl_values + m_capacity,
                     m_hashtbl_values);
   }
+
+  /**
+   * @note If called on host using managed allocator, then caller should ensure
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Same as previous changes.

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 10]
```diff
@@ -149,18 +157,53 @@ class concurrent_unordered_map {
...
   }
 
+  /**
+   * @note If called on host using managed allocator, then caller should ensure
+   * the stream is synchronized before it is accessed from the host.
+   *
+   * @note When called in a device code, user should make sure that it should
+   * either be running on the same stream as create(), or the accessing stream
+   * should be appropriately synchronized with the creating stream.
+   **/
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]:
```suggestion
   * @returns iterator to the first element in the map.
   **/
```

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 11]
```diff
@@ -149,18 +157,53 @@ class concurrent_unordered_map {
...
   __device__ const_iterator begin() const {
     return const_iterator(m_hashtbl_values, m_hashtbl_values + m_capacity,
                           m_hashtbl_values);
   }
+
+  /**
+   * @note If called on host using managed allocator, then caller should ensure
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Same as previous changes.

File: cpp/src/hash/concurrent_unordered_multimap.cuh [hunk 12]
```diff
@@ -35,6 +35,10 @@
 /**
  * Does support concurrent insert, but not concurrent insert and probping.
  *
+ * @note User is responsible for following stream semantics:
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_multimap.cuh]: Same suggestions as for map.

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 13]
```diff
@@ -149,18 +157,53 @@ class concurrent_unordered_map {
...
+   * either be running on the same stream as create(), or the accessing stream
+   * should be appropriately synchronized with the creating stream.
+   **/
   __device__ iterator end() {
     return iterator(m_hashtbl_values, m_hashtbl_values + m_capacity,
                     m_hashtbl_values + m_capacity);
   }
+
+  /**
+   * @note If called on host using managed allocator, then caller should ensure
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Same as previous changes.

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 14]
```diff
@@ -149,18 +157,53 @@ class concurrent_unordered_map {
         deleter};
   }
 
+  /**
+   * @note If called on host using managed allocator, then caller should ensure
+   * the stream is synchronized before it is accessed from the host.
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]:
```suggestion
```

File: cpp/src/hash/concurrent_unordered_multimap.cuh [hunk 15]
```diff
@@ -114,18 +122,53 @@ class concurrent_unordered_multimap {
     delete this;
   }
 
+  /**
+   * @note If called on host using managed allocator, then caller should ensure
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_multimap.cuh]:
```suggestion
   * @note When using the managed allocator, host code that calls this function should ensure
```

File: cpp/src/hash/concurrent_unordered_multimap.cuh [hunk 16]
```diff
@@ -114,18 +122,53 @@ class concurrent_unordered_multimap {
...
+   * the stream is synchronized before it is accessed from the host.
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_multimap.cuh]:
```suggestion
   * the stream used for `create()` is appropriately synchronized.
```

Comment (harrism) [reply cpp/src/hash/concurrent_unordered_multimap.cuh]: Also add brief and returns docs.

File: cpp/src/hash/concurrent_unordered_multimap.cuh [hunk 17]
```diff
@@ -114,18 +122,53 @@ class concurrent_unordered_multimap {
...
+   * either be running on the same stream as create(), or the accessing stream
+   * should be appropriately synchronized with the creating stream.
+   **/
   __host__ __device__ iterator begin() {
     return iterator(m_hashtbl_values, m_hashtbl_values + m_hashtbl_size,
                     m_hashtbl_values);
   }
+
+  /**
+   * @note If called on host using managed allocator, then caller should ensure
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_multimap.cuh]: Same as previous suggestions.

File: cpp/src/hash/concurrent_unordered_multimap.cuh [hunk 18]
```diff
@@ -114,18 +122,53 @@ class concurrent_unordered_multimap {
...
+   * either be running on the same stream as create(), or the accessing stream
+   * should be appropriately synchronized with the creating stream.
+   **/
   __host__ __device__ const_iterator begin() const {
     return const_iterator(m_hashtbl_values, m_hashtbl_values + m_hashtbl_size,
                           m_hashtbl_values);
   }
+
+  /**
+   * @note If called on host using managed allocator, then caller should ensure
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_multimap.cuh]: Same as previous suggestions.

File: cpp/src/hash/concurrent_unordered_multimap.cuh [hunk 19]
```diff
@@ -114,18 +122,53 @@ class concurrent_unordered_multimap {
...
+   * either be running on the same stream as create(), or the accessing stream
+   * should be appropriately synchronized with the creating stream.
+   **/
   __host__ __device__ iterator end() {
     return iterator(m_hashtbl_values, m_hashtbl_values + m_hashtbl_size,
                     m_hashtbl_values + m_hashtbl_size);
   }
+
+  /**
+   * @note If called on host using managed allocator, then caller should ensure
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_multimap.cuh]: Same as previous suggestions.

File: cpp/src/hash/concurrent_unordered_map.cuh (hunk 4 above)
Comment (rgsl888prabhu) [reply cpp/src/hash/concurrent_unordered_map.cuh]: `add constructor that takes pointer to hash_table to avoid allocations` still needs to be done

**Pull request merged**